In [78]:
# IMPORT LIBRARIES

In [79]:
import pandas as pd
import sqlite3

In [80]:
# CONNECT DATABASE

In [81]:
conn = sqlite3.connect('ipl_data.db')
print("Database connected successfully")

Database connected successfully


In [82]:
# TOTAL MATCHES PER SEASON

In [83]:
q1 = pd.read_sql_query(" select season, count(*) as matches_played from matches where team1 = 'Royal Challengers Bengaluru' or team2 = 'Royal Challengers Bengaluru' group by season ",conn)
print("Matches per season:")
print(q1)

Matches per season:
     season  matches_played
0   2007/08              14
1      2009              16
2   2009/10              16
3      2011              16
4      2012              15
5      2013              16
6      2014              14
7      2015              16
8      2016              16
9      2017              13
10     2018              14
11     2019              14
12  2020/21              15
13     2021              15
14     2022              16
15     2023              14
16     2024              15


In [84]:
# WINS PER SEASON

In [85]:
q2 = pd.read_sql_query("""
    SELECT 
        season,
        COUNT(*) as matches_won
    FROM matches 
    WHERE winner LIKE '%Royal Challengers%' 
    GROUP BY season, winner
    ORDER BY season
""", conn)
print("Wins per season:")
display(q2)

Wins per season:


,season,matches_won
0,2007/08,4
1,2009,9
2,2009/10,8
3,2011,10
4,2012,8
5,2013,9
6,2014,5
7,2015,8
8,2016,9
9,2017,3


In [86]:
# COMBINED: MATCHES + WINS PER SEASON (CASE WHEN)

In [87]:
q3 = pd.read_sql_query("""
    SELECT 
        season,
        COUNT(*) as matches_played,
        COUNT(CASE WHEN winner LIKE '%Royal Challengers%' THEN 1 END) as matches_won
    FROM matches 
    WHERE team1 = 'Royal Challengers Bengaluru' or team2 = 'Royal Challengers Bengaluru'
    GROUP BY season
    ORDER BY season
""", conn)
print("Combined season data:")
display(q3)

Combined season data:


,season,matches_played,matches_won
0,2007/08,14,4
1,2009,16,9
2,2009/10,16,8
3,2011,16,10
4,2012,15,8
5,2013,16,9
6,2014,14,5
7,2015,16,8
8,2016,16,9
9,2017,13,3


In [88]:
# WIN PERCENTAGE PER SEASON

In [89]:
q4 = pd.read_sql_query("""
    SELECT 
        season,
        COUNT(*) as matches_played,
        COUNT(CASE WHEN winner LIKE '%Royal Challengers%' THEN 1 END) as matches_won,
        ROUND(COUNT(CASE WHEN winner LIKE '%Royal Challengers%' THEN 1 END) * 100.0 / COUNT(*), 1) as win_pct
    FROM matches 
    WHERE team1 = 'Royal Challengers Bengaluru' or team2 = 'Royal Challengers Bengaluru'
    GROUP BY season
    ORDER BY season
""", conn)
print("Win percentage per season:")
display(q4)

Win percentage per season:


,season,matches_played,matches_won,win_pct
0,2007/08,14,4,28.6
1,2009,16,9,56.3
2,2009/10,16,8,50.0
3,2011,16,10,62.5
4,2012,15,8,53.3
5,2013,16,9,56.3
6,2014,14,5,35.7
7,2015,16,8,50.0
8,2016,16,9,56.3
9,2017,13,3,23.1


In [90]:
# BEST SEASON (HIGHEST WIN%) 

In [91]:
q5 = """
SELECT 
    season,
    COUNT(*) AS total_matches,
    COUNT(CASE WHEN winner like '%Royal Challengers%' THEN 1 END) AS wins,
    ROUND(COUNT(CASE WHEN winner like '%Royal Challengers%' THEN 1 END) * 100.0 / COUNT(*), 1) AS win_pct
FROM matches
WHERE team1 = 'Royal Challengers Bengaluru'
   OR team2 = 'Royal Challengers Bengaluru'
GROUP BY season
ORDER BY win_pct DESC
LIMIT 1
"""

best_season = pd.read_sql_query(q5, conn)
print("Best season:")
print(best_season)

Best season:
  season  total_matches  wins  win_pct
0   2011             16    10     62.5


In [92]:
# WORST SEASON (LOWEST WIN%)

In [93]:
q6 = """
SELECT 
    season,
    COUNT(*) AS total_matches,
    COUNT(CASE WHEN winner like '%Royal Challengers%' THEN 1 END) AS wins,
    ROUND(COUNT(CASE WHEN winner like '%Royal Challengers%' THEN 1 END) * 100.0 / COUNT(*), 1) AS win_pct
FROM matches
WHERE team1 = 'Royal Challengers Bengaluru'
   OR team2 = 'Royal Challengers Bengaluru'
GROUP BY season
ORDER BY win_pct ASC
LIMIT 1
"""

worst_season = pd.read_sql_query(q6, conn)
print("Worst season:")
print(worst_season)

Worst season:
  season  total_matches  wins  win_pct
0   2017             13     3     23.1


In [94]:
# OVERALL WIN PERCENTAGE

In [95]:
q7 = """
SELECT 
    COUNT(*) AS total_matches,
    COUNT(CASE WHEN winner like '%Royal Challengers%' THEN 1 END) AS wins,
    ROUND(COUNT(CASE WHEN winner like '%Royal Challengers%' THEN 1 END) * 100.0 / COUNT(*), 1) AS win_pct
FROM matches
WHERE team1 = 'Royal Challengers Bengaluru'
   OR team2 = 'Royal Challengers Bengaluru'
"""

overall = pd.read_sql_query(q7, conn)
print("Overall RCB record:")
print(overall)


Overall RCB record:
   total_matches  wins  win_pct
0            255   123     48.2


In [96]:
# CLOSE CONNECTION

In [97]:
conn.close()
print("Connection closed.")

Connection closed.


## Day 3 Summary: Aggregation & Win Percentage

Completed
- Calculated matches and wins per season in a single query using CASE WHEN
- Computed win percentage per season with ROUND()
- Identified best and worst seasons using ORDER BY + LIMIT
- Verified overall win%: 48.2% — matches Pandas result

Key Queries Learned
- COUNT with CASE WHEN for conditional counting
- ROUND() for decimal formatting
- Arithmetic in SELECT (wins * 100.0 / total)
- ORDER BY ... DESC LIMIT 1 for top/bottom results

Verification
✅ Overall win%: 48.2% — matches Pandas EDA  
✅ Total matches: 255, Wins: 123, Losses: 132

Next
Day 4 — CASE Statements: Categorizing match outcomes and player performance